controllers/openstackcluster_controller.go

```diff
@@ -189,17 +189,17 @@ func (r *OpenStackClusterReconciler) reconcileDelete(ctx context.Context, log lo
 		r.Recorder.Eventf(openStackCluster, corev1.EventTypeNormal, "SuccessfulDeleteSecurityGroup", "Deleted security group %s with id %s", controlPlaneSecGroup.Name, controlPlaneSecGroup.ID)
 	}
 
-	if router := openStackCluster.Status.Network.Router; router != nil {
-		log.Info("Deleting router", "name", router.Name)
-		if err = networkingService.DeleteRouter(openStackCluster.Status.Network); err != nil {
-			return ctrl.Result{}, errors.Errorf("failed to delete router: %v", err)
-		}
-		r.Recorder.Eventf(openStackCluster, corev1.EventTypeNormal, "SuccessfulDeleteRouter", "Deleted router %s with id %s", router.Name, router.ID)
-		log.Info("OpenStack router deleted successfully")
-	}
-
 	// if NodeCIDR was not set, no network was created.
 	if network := openStackCluster.Status.Network; network != nil && openStackCluster.Spec.NodeCIDR != "" {
+		if router := network.Router; router != nil {
```

Comment:
Even if openStackCluster.Spec.NodeCIDR == "", router should be deleted;

```go
if network := openStackCluster.Status.Network; network != nil {
  if router := network.Router; router != nil {
    // delete router
  }
  if openStackCluster.Spec.NodeCIDR != "" {
    // delete network
  }
}
```

Reply:
From what I understood ([Ref](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/controllers/openstackcluster_controller.go#L402)), the router is created only when `openStackCluster.Spec.NodeCIDR != ""`. Thus we should be deleting the router only in that scenario.
Please correct me if my understanding is wrong.

Reply:
I was wrong, you are right. 
I understand you fix to guard openStackCluster.Status.Network, and openStackCluster.Spec.NodeCIDR != "".